File: app/python/ml_tdd/script/train_boston.py
"""Boston 住宅価格予測モデルの訓練スクリプト."""

import os
from pathlib import Path

from sklearn.model_selection import train_test_split

from src.ml.boston_predictor import BostonPredictor


def main() -> None:
    """メイン処理."""
    # プロジェクトルートディレクトリを取得
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    data_path = project_root / "data" / "Boston.csv"
    model_dir = project_root / "model"

    # モデル保存ディレクトリが存在しない場合は作成
    os.makedirs(model_dir, exist_ok=True)

    # 予測器の作成
    predictor = BostonPredictor()
    print("BostonPredictor を作成しました")

    # データの読み込み
    df = predictor.load_data(str(data_path))
    print(f"データを読み込みました: {len(df)} サンプル")

    # CRIME 列のダミー変数化
    df = predictor._encode_crime(df)
    print("CRIME 列をダミー変数化しました")

    # 欠損値の補完
    df = predictor._fill_missing_values(df, fit=True)
    print("欠損値を補完しました")

    # 外れ値の除外
    df = predictor._remove_outliers(df)
    print(f"外れ値を除外しました: {len(df)} サンプル")

    # 特徴量と目的変数の分離
    feature_columns = ["RM", "LSTAT", "PTRATIO"]
    X = df[feature_columns]
    y = df[["PRICE"]]

    # 訓練データとテストデータに分割
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=0
    )
    print(f"訓練データ: {len(X_train)} サンプル")
    print(f"テストデータ: {len(X_test)} サンプル")

    # 特徴量エンジニアリング
    X_train_eng = predictor.feature_engineering(X_train)
    X_test_eng = predictor.feature_engineering(X_test)
    print(f"特徴量エンジニアリング完了: {X_train_eng.shape[1]} 特徴量")

    # 特徴量の標準化
    X_train_scaled = predictor.standardize_features(X_train_eng, fit=True)
    X_test_scaled = predictor.standardize_features(X_test_eng, fit=False)
    print("特徴量を標準化しました")

    # 目的変数の標準化
    y_train_scaled = predictor.standardize_target(y_train, fit=True)
    y_test_scaled = predictor.standardize_target(y_test, fit=False)
    print("目的変数を標準化しました")

    # モデルの訓練
    predictor.train(X_train_scaled, y_train_scaled)
    print("\nモデルの訓練が完了しました")

    # モデルの評価（標準化されたデータで）
    score = predictor.evaluate(X_test_scaled, y_test_scaled)
    print("\n[モデルの評価]")
    print(f"  決定係数（R^2）: {score:.4f}")

    # モデルとスケーラーの保存
    model_path = model_dir / "boston_model.pkl"
    scaler_x_path = model_dir / "boston_scaler_X.pkl"
    scaler_y_path = model_dir / "boston_scaler_y.pkl"

    predictor.save_models(
        str(model_path),
        str(scaler_x_path),
        str(scaler_y_path),
    )
    print("\nモデルとスケーラーを保存しました")
    print(f"  - {model_path}")
    print(f"  - {scaler_x_path}")
    print(f"  - {scaler_y_path}")

    # 予測例
    print("\n[予測例]")
    sample_test = X_test_eng.iloc[:5]
    sample_test_scaled = predictor.standardize_features(sample_test, fit=False)
    predictions_scaled = predictor.predict(sample_test_scaled)

    # 予測結果を元のスケールに戻す
    predictions = predictor.inverse_transform_prediction(
        predictions_scaled.reshape(-1, 1)
    )

    for i in range(len(sample_test)):
        actual = y_test.iloc[i, 0]
        predicted = predictions[i, 0]
        error = abs(actual - predicted)

        print(f"\nサンプル {i + 1}:")
        print(
            f"  特徴量: RM={X_test.iloc[i]['RM']:.2f}, "
            f"LSTAT={X_test.iloc[i]['LSTAT']:.2f}, "
            f"PTRATIO={X_test.iloc[i]['PTRATIO']:.2f}"
        )
        print(f"  実際の価格: ${actual:.2f}K")
        print(f"  予測価格:   ${predicted:.2f}K")
        print(f"  誤差:       ${error:.2f}K")


if __name__ == "__main__":
    main()

File: app/typescript/scripts/train-boston.ts
import { BostonPredictor } from '../src/models/BostonPredictor';
import * as path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, '..');

async function main(): Promise<void> {
  console.log('=== Boston 住宅価格予測モデルの訓練 ===\n');

  const predictor = new BostonPredictor();
  const dataPath = path.join(projectRoot, 'data/boston.csv');

  // データ読み込み
  console.log('📂 データを読み込んでいます...');
  let df = await predictor.loadData(dataPath);
  console.log(`✓ データ読み込み完了: ${df.count()} 件\n`);

  // データの前処理パイプライン
  console.log('🔧 データの前処理を実行中...\n');

  // 1. CRIME 列のダミー変数化
  console.log('  1. CRIME 列をダミー変数化');
  df = (predictor as any).encodeCrime(df);

  // 2. 欠損値補完（訓練データで fit）
  console.log('  2. 欠損値を補完');
  df = (predictor as any).fillMissingValues(df, true);

  // 3. 外れ値除外
  console.log('  3. 外れ値を除外');
  df = (predictor as any).removeOutliers(df);
  console.log(`     → 除外後のデータ数: ${df.count()} 件\n`);

  // 特徴量と目的変数の分割
  const featureColumns = ['RM', 'LSTAT', 'PTRATIO'];
  let X = df.subset(featureColumns);
  const y = df.getSeries('PRICE');

  // 4. 特徴量エンジニアリング
  console.log('⚙️ 特徴量エンジニアリング実行中...');
  X = predictor.featureEngineering(X);
  console.log(`  → 特徴量数: ${X.getColumnNames().length} 個`);
  console.log(`     ${X.getColumnNames().join(', ')}\n`);

  // データを配列に変換
  const XArray = X.toArray().map((row) =>
    X.getColumnNames().map((col) => Number(row[col]))
  );
  const yArray = y.toArray().map((val) => Number(val));

  // 訓練データとテストデータに分割（80/20）
  const totalSize = XArray.length;
  const trainSize = Math.floor(totalSize * 0.8);

  const XTrain = XArray.slice(0, trainSize);
  const yTrain = yArray.slice(0, trainSize);
  const XTest = XArray.slice(trainSize);
  const yTest = yArray.slice(trainSize);

  console.log('📊 データ分割:');
  console.log(`  訓練データ: ${XTrain.length} 件`);
  console.log(`  テストデータ: ${XTest.length} 件\n`);

  // 5. データ標準化
  console.log('📐 データを標準化中...');
  const XTrainScaled = predictor.standardizeFeatures(XTrain, true);
  const yTrainScaled = predictor.standardizeTarget(yTrain, true);
  const XTestScaled = predictor.standardizeFeatures(XTest, false);
  console.log('✓ 標準化完了\n');

  // 6. モデルの訓練
  console.log('🤖 モデルを訓練中...');
  predictor.train(XTrainScaled, yTrainScaled);
  console.log('✓ 訓練完了\n');

  // 7. モデルの評価
  console.log('📈 モデルを評価中...');

  // 訓練データでの評価
  const yTrainPredScaled = predictor.predict(XTrainScaled);
  const yTrainPred = predictor.inverseTransformPrediction(yTrainPredScaled);

  const trainMean = yTrain.reduce((a, b) => a + b, 0) / yTrain.length;
  const trainSST = yTrain.reduce((sum, y) => sum + (y - trainMean) ** 2, 0);
  const trainSSR = yTrain.reduce(
    (sum, y, i) => sum + (y - yTrainPred[i]) ** 2,
    0
  );
  const trainR2 = 1 - trainSSR / trainSST;

  console.log(`  訓練データ R²: ${trainR2.toFixed(4)}`);

  // テストデータでの評価
  const yTestPredScaled = predictor.predict(XTestScaled);
  const yTestPred = predictor.inverseTransformPrediction(yTestPredScaled);

  const testMean = yTest.reduce((a, b) => a + b, 0) / yTest.length;
  const testSST = yTest.reduce((sum, y) => sum + (y - testMean) ** 2, 0);
  const testSSR = yTest.reduce((sum, y, i) => sum + (y - yTestPred[i]) ** 2, 0);
  const testR2 = 1 - testSSR / testSST;

  console.log(`  テストデータ R²: ${testR2.toFixed(4)}\n`);

  // 8. モデルとスケーラーの保存
  console.log('💾 モデルとスケーラーを保存中...');
  const modelPath = path.join(projectRoot, 'models/boston_model.json');
  const scalerXPath = path.join(projectRoot, 'models/boston_scalerX.json');
  const scalerYPath = path.join(projectRoot, 'models/boston_scalerY.json');
  const singleFilePath = path.join(projectRoot, 'models/boston_predictor.json');

  await predictor.saveModels(modelPath, scalerXPath, scalerYPath);
  await predictor.save(singleFilePath);
  console.log(`✓ モデルを保存しました:`);
  console.log(`  - ${modelPath}`);
  console.log(`  - ${scalerXPath}`);
  console.log(`  - ${scalerYPath}`);
  console.log(`  - ${singleFilePath} (単一ファイル形式)\n`);

  // 予測例の表示
  console.log('🏠 予測例:');
  console.log('─'.repeat(70));
  console.log(
    'RM      LSTAT   PTRATIO   | 実際の価格  予測価格  差分'
  );
  console.log('─'.repeat(70));

  for (let i = 0; i < Math.min(10, XTest.length); i++) {
    const features = XTest[i];
    const actual = yTest[i];
    const predicted = yTestPred[i];
    const diff = actual - predicted;

    console.log(
      `${features[0].toFixed(1).padStart(6)}  ` +
        `${features[1].toFixed(1).padStart(6)}  ` +
        `${features[2].toFixed(1).padStart(6)}    | ` +
        `$${(actual * 1000).toFixed(0).padStart(6)}  ` +
        `$${(predicted * 1000).toFixed(0).padStart(6)}  ` +
        `${diff > 0 ? '+' : ''}${(diff * 1000).toFixed(0)}`
    );
  }
  console.log('─'.repeat(70));

  console.log('\n=== 訓練完了 ===');
}

main().catch((error) => {
  console.error('エラーが発生しました:', error);
  process.exit(1);
});
